### ISSUE TITLE
SelectorRow documentation is corrupt

### ISSUE THREAD

**Issue by fhchina** [xmartlabs/Eureka#1282]
How to develop a Custom Presenter row in Eureka 4.x? The offocial documentation is still 

public final class CustomPushRow<T: Equatable>: SelectorRow<PushSelectorCell<T>, SelectorViewController<T>>, RowType {

```swift
    public required init(tag: String?) {
        super.init(tag: tag)
        presentationMode = .show(controllerProvider: ControllerProvider.callback {
            return SelectorViewController<T>(){ _ in }
            }, onDismiss: { vc in
                _ = vc.navigationController?.popViewController(animated: true)
        })
    }
}
```

but this couldn't get compiled.

**Comment by kiwiswift** [comment]
I'm facing a similar problem. After following the hints from issue #776, I'm still getting the error messages below:
Type 'CustomPushRow' does not conform to protocol 'TypedRowType'
Type 'CustomPushRow' does not conform to protocol 'Taggable'
Type 'CustomPushRow' does not conform to protocol 'RowType'
Type 'CustomPushRow' does not conform to protocol 'BaseRowType'
Generic type 'SelectorRow' specialized with too many type parameters (got 2, but expected 1)

My code:
`open class CustomPushCell : Cell<Int>, CellType{
    
    required public init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
}

class CustomViewController: UIViewController, TypedRowControllerType{
    public var row: RowOf<Int>!
    public var onDismissCallback : ((UIViewController) -> ())?
}

public final class CustomPushRow: SelectorRow<CustomPushCell, CustomViewController>, RowType {
    
    public required init(tag: String?) {
        super.init(tag: tag)
        presentationMode = .presentModally(controllerProvider: ControllerProvider.callback {
            let vc = CustomViewController()
            vc.row = self
            return vc
            }, onDismiss: { vc in
                _ = vc.navigationController?.popViewController(animated: true)
        })
    }
}

`

**Comment by chappejw** [comment]
In your code try removing CustomViewController as in the following code

```
public final class CustomPushRow: SelectorRow<CustomPushCell>, 
RowType {

    public required init(tag: String?) {
        super.init(tag: tag)
        presentationMode = .presentModally(controllerProvider: ControllerProvider.callback {
            let vc = CustomViewController()
            vc.row = self
            return vc
            }, onDismiss: { vc in
                _ = vc.navigationController?.popViewController(animated: true)
        })
    }
}
```

You should get through the initial error, but then you may find another issue in regards to a type mismatch. This did work on some similar methods for me, but currently I am blocked by an error:

I'm not sure yet what is wrong with my implementation. On the line with presentationMode = .show(... I get this error:

Cannot convert value of type 'ControllerProvider<MapViewController>' to expected argument type 'ControllerProvider<_>'

If anyone has some insight to what I am doing wrong that would be awesome.

```
public final class GeolocationRow : SelectorRow<PushSelectorCell<CLLocation>>, RowType {
    public required init(tag: String?) {
        super.init(tag: tag)
        presentationMode = .show(controllerProvider: ControllerProvider.callback { return MapViewController(){ _ in } }, onDismiss: { vc in
            let _ = vc.navigationController?.popViewController(animated: true) })
        displayValueFor = {
            guard let location = $0 else { return "" }
            let fmt = NumberFormatter()
            fmt.maximumFractionDigits = 4
            fmt.minimumFractionDigits = 4
            let latitude = fmt.string(from: NSNumber(value: location.coordinate.latitude))!
            let longitude = fmt.string(from: NSNumber(value: location.coordinate.longitude))!
            return  "\(latitude), \(longitude)"
        }
    }
}
```

**Comment by kiwiswift** [comment]
Thanks @chappejw, I tried that before and got the same error you are getting.

**Comment by mspasov** [comment]
So, you're getting this, because with recent changes, SelectorRow is more specialized (select from list of options) and the functionality just to present a custom TypedRowControllerType is now lost. Try using This row as base, should work:

```
open class CustomViewControllerRow<Cell: CellType, VCType: TypedRowControllerType>: OptionsRow<Cell>, PresenterRowType where Cell: BaseCell, VCType: UIViewController, VCType.RowValue == Cell.Value {

    public var presentationMode: PresentationMode<VCType>?
    public var onPresentCallback: ((FormViewController, VCType) -> Void)?
    public typealias PresentedControllerType = VCType
    
    public typealias PresenterRow = VCType
    
    required public init(tag: String?) {
        super.init(tag: tag)
    }
    open override func customDidSelect() {
        super.customDidSelect()
        guard let presentationMode = presentationMode, !isDisabled else { return }
        if let controller = presentationMode.makeController() {
            controller.row = self
            controller.title = selectorTitle ?? controller.title
            onPresentCallback?(cell.formViewController()!, controller)
            presentationMode.present(controller, row: self, presentingController: self.cell.formViewController()!)
        } else {
            presentationMode.present(nil, row: self, presentingController: self.cell.formViewController()!)
        }
    }
}
```

ps. It will be a great idea to have a such a row in Eureka.

**Comment by funkenstrahlen** [comment]
I was happy to find this post. I also had this problem. I think the Readme should be updated soon!

**Comment by eliburke** [comment]
Huge kudos to @mspasov 
Added the code snipped, changed the inheritance, good to go.

**Comment by guillianbalisi** [comment]
Can someone provide a final solution including the implementation of the CustomPushRow?

**Comment by kohdesmond** [comment]
@mspasov how do i fix my code from swift 3 to 4? any pointers are greatly appreciated!

**Comment by damiancesar** [comment]
Can someone help me implement the base row @mspasov provided? 

How would I add it to the form? 

Do I need to create a custom cell? What if I want the same cell style as the "PushRow"?

**Comment by damiancesar** [comment]
If anyone needs help implementing @mspasov base row, here's how I've done it:

```
final class PresenterRow: _PresenterRow<PushSelectorCell<String>, UniversityStructureTableViewController>, RowType { 
    // Implementation here
}
```

```
class CustomFormViewController: FormViewController {
    override func viewDidLoad() {
        super.viewDidLoad()
        
        form +++ Section()
            <<< PresenterRow() { row in
            row.presentationMode = .show(controllerProvider: ControllerProvider.callback {
                let storyboard = UIStoryboard(name: "UniversityStructure", bundle: .main)
                let controller = storyboard.instantiateInitialViewController() as! UniversityStructureTableViewController
                controller.delegate = row
                return controller
                }, onDismiss: { vc in
                    _ = vc.navigationController?.popViewController(animated: true)
            })
        }
    }
}
```

**Comment by aadjemonkeyrock** [comment]
@damiancesar I copied your example. but can't get it to compile with Swift 4.1. 

<img width="925" alt="tripviewcontroller_swift_ _edited" src="https://user-images.githubusercontent.com/956176/44752284-d53f6080-ab1a-11e8-931d-3aff3f7c4daf.png">

Any help would be appreciated.

**Comment by aadjemonkeyrock** [comment]
@damiancesar Unfortunately not, I actually stopped the project for a while.. ;-)

**Comment by cesaredecal** [comment]
Hey @damiancesar it's not working on my end either. To test it, simply copy your code into a new Swift file.

**Comment by adamdahan** [comment]
This is retarded. Why the in the world would it be so complicated to push from one vc to another. -10000.

**Comment by adamdahan** [comment]
Just use a custom callback inside the onCellSelection -> Pass the payload and move the controller yourself. Don't b crazy.

**Comment by adamdahan** [comment]
`for address in addresses {
            self.form.sectionBy(tag: "results")! <<< ButtonRow{ row in
                row.title = address
                row.cell.textLabel?.textAlignment = .left
                row.cell.height = { 70.0 }
                row.onCellSelection({ (button, row) in
                    // push to view controller with your data
                    // in my case "address"
                })
                row.cellUpdate({ (button, row) in
                    row.cell.textLabel?.textAlignment = .left
                })
            }
        }
`
